src/Compiler/Driver/ParseAndCheckInputs.fs

```diff
@@ -1534,13 +1572,418 @@ let CheckMultipleInputsInParallel
+
+                        // backed impl files must not add results as there are already results from .fsi files
+                        //let fsTcState = if backed then tcState else fsTcState
+
+                        let partialResult = tcEnvAtEnd, topAttrs, Some implFile, ccuSigForFile
+
+                        let tcState =
+                            { fsTcState with
+                                tcsCreatesGeneratedProvidedTypes =
+                                    fsTcState.tcsCreatesGeneratedProvidedTypes || createsGeneratedProvidedTypes
+                            }
+
+                        // printfn $"Finished applying Impl {file.FileName}"
+                        partialResult, tcState
+
+        with e ->
+            errorRecovery e range0
+            return fun tcState -> (tcState.TcEnvFromSignatures, EmptyTopAttrs, None, tcState.tcsCcuSig), tcState
+    }
+
+let AddSignatureResultToTcImplEnv (tcImports: TcImports, tcGlobals, prefixPathOpt, tcSink, tcState, input: ParsedInput) =
+    let qualNameOfFile = input.QualifiedName
+    let rootSigOpt = tcState.tcsRootSigs.TryFind qualNameOfFile
+
+    match rootSigOpt with
+    | None -> failwithf $"No signature data was found for %s{input.FileName}"
+    | Some rootSig ->
+        fun (tcState: TcState) ->
+            let amap = tcImports.GetImportMap()
+
+            // Add the results of type checking the signature file to the TcEnv of implementation files.
+            let ccuSigForFile, tcState =
+                AddCheckResultsToTcState
+                    (tcGlobals, amap, true, prefixPathOpt, tcSink, tcState.tcsTcImplEnv, qualNameOfFile, rootSig)
+                    tcState
+
+            // This partial result will be discarded in the end of the graph resolution.
+            let partialResult: PartialResult =
+                tcState.tcsTcSigEnv, EmptyTopAttrs, None, ccuSigForFile
+
+            partialResult, tcState
+
+/// Constructs a file dependency graph and type-checks the files in parallel where possible.
+let CheckMultipleInputsUsingGraphMode
+    ((ctok, checkForErrors, tcConfig: TcConfig, tcImports: TcImports, tcGlobals, prefixPathOpt, tcState, eagerFormat, inputs): 'a * (unit -> bool) * TcConfig * TcImports * TcGlobals * LongIdent option * TcState * (PhasedDiagnostic -> PhasedDiagnostic) * ParsedInput list)
+    : FinalFileResult list * TcState =
+    use cts = new CancellationTokenSource()
+
+    let sourceFiles: FileWithAST array =
+        inputs
+        |> List.toArray
+        |> Array.mapi (fun idx (input: ParsedInput) ->
+            {
+                Idx = idx
+                File = input.FileName
+                AST = input
+            })
+
+    let filePairs = FilePairMap(sourceFiles)
+
+    let graph =
+        let graph = DependencyResolution.mkGraph filePairs sourceFiles
+
+        if not tcConfig.compilingFSharpCore then
+            graph
+        else
+            // Every single file should depend on prim-types-prelude.fsi
```

Comment:
@safesparrow when testing this with `FSharp.Core` I noticed that type-checking was failing.
The very first file (`prim-types-prelude.fsi`) in the project needs to be referenced by all the others.

This file has the aliases for all the built-in types:

```fsharp
// Basic F# type definitions, functions and operators.
namespace Microsoft.FSharp.Core

    open System

    /// <summary>An abbreviation for the CLI type <see cref="T:System.Object"/>.</summary>
    ///
    /// <category>Basic Types</category>
    type obj = System.Object
```

Very concretely, the links were missing for some of the generated files:
```
artifacts\obj\FSharp.Core\Release\netstandard2.1\buildproperties.fs
artifacts\obj\FSharp.Core\Release\netstandard2.1\.NETStandard,Version=v2.1.AssemblyAttributes.fs
artifacts\obj\FSharp.Core\Release\netstandard2.1\FSharp.Core.AssemblyInfo.fs
```

When looking at these files, AST-wise there is no reason to link them.
`buildproperties.fs` looks like this:
```fsharp
//  <auto-generated >
//    <Generated by the FSharp WriteCodeFragment class./>
//  </auto-generated/>
//
module internal FSharp.BuildProperties
let fsProductVersion = "12.0.5.0"
let fsLanguageVersion = "7.0"
```

This extra bit of code resolves the problem and made the CI leg pass.

Reply:
Why the check for Seq.isEmpty ? If they had a link between each other (e.g. AssemblyInfo having a link to AssemblyAttributes), edge to the 0th node should still be added or?

Reply:
Those three files had no links at all. All the other files had a direct link to `prim-types-prelude.fs`.

Reply:
I know they do not have it now, I would just see a "set-based append of 0" (add 0 to the set of edges represented by the array) more clear in communicating the intent => the intent is to ensure that each file has an edge towards the first one.

Reply:
Do we understand why the link is needed but not inferrable from the AST? What exactly is used from that file?

Reply:
From the project aspect, this is what I see as outstanding, the CompileBefore element:

<img width="362" alt="image" src="https://user-images.githubusercontent.com/46543583/210753026-120cb8f0-8ee9-4a32-aabf-2f7b70655a6a.png">

Reply:
Which forces that file to be forcibly compiled befre the 3 generated files.
I think the reason has to do with it providing shortcuts for the most basic of types - even using a string literal might need it (guessing)

Reply:
We should get to the bottom of this. Is the link needed because Microsoft.FSharp.Core namespace is special? Is it on the list of implicitly opened namespaces? There was a discussion about such namespaces a while ago on Slack.

What does CompileBefore do? Does it just change file ordering? As I understand we have the right order of files, but the problem is we see no reason to create a link.

Reply:
> Is the link needed because Microsoft.FSharp.Core namespace is special?

* FSharp.Core assembly has an attribute that makes it automatically imported.
* There is logic that allows using `FSharp.Core` instead of `Microsoft.FSharp.Core`
* Maybe there's something else enabled when `TcConfigBuilder.compilingFSharpCore` is set

> What does CompileBefore do? Does it just change file ordering? 

Yes, it only changes the order. FCS/fsc requests expect the correct order of sources, so it's a task for an IDE/MSBuild to order them properly.

Reply:
I don't think there is any issue with compileafter or compilebefore - those are handled by msbuild. What is the issue is the implicit opening of certain namespaces in the fsharp.core project, which is/was not encoded in the dependency resolution mechanism.

Reply:
I still can't put my finger on this problem but I'm quite sure it is very isolated to compiling F# Core.

```
prim-types-prelude.fsi
prim-types-prelude.fs
C:\Users\nojaf\Projects\safesparrow-fsharp\artifacts\obj\FSharp.Core\Debug\netstandard2.0\FSCore.fs
C:\Users\nojaf\Projects\safesparrow-fsharp\artifacts\obj\FSharp.Core\Debug\netstandard2.0\buildproperties.fs
C:\Users\nojaf\Projects\safesparrow-fsharp\artifacts\obj\FSharp.Core\Debug\netstandard2.0\FSharp.Core.AssemblyInfo.fs
prim-types.fsi
prim-types.fs
```

Every file that is between `prim-types-prelude.fsi` and `prim-types.fsi` needs a link to `prim-types-prelude.fsi`.

Reply:
What would make sense to me is if some part of type-checking assumed that definitions from `prim-types-prelude` always exist - they do when compiling any other project (as `Microsoft.FSharp.core` is auto-opened when `FSharp.Core` project is referenced), and in the sequential approach they exist in all files after `prim-types-prelude.*`.

The only thing I can't figure out though is how does type-checking of these two files `prim-types.prelude.*` work then?
Why does `buildproperties.fs` not type-check without it if it doesn't need anything from it explicitly?
Is there some hidden code that treats the prelude files differently and thus makes them type-check fine?

Reply:
I was wrong - `CompileBefore` changes more than order - compiling FSharp.Core with just two files `prim-types-prelude.*` yields different results (errors) when using `Compile` vs `CompileBefore`. I'll have a look at how it's different.